extensions/mqtt/processors/AbstractMQTTProcessor.cpp

```diff
@@ -15,147 +15,167 @@
       }
-      value = "";
-      if (context->getProperty(SecurityPrivateKey.getName(), value) && !value.empty()) {
-        logger_->log_debug("AbstractMQTTProcessor: privateKey [%s]", value);
-        securityPrivateKey_ = value;
-        sslopts_.privateKey = securityPrivateKey_.c_str();
+      if (auto value = context->getProperty(SecurityPrivateKey)) {
+        logger_->log_debug("AbstractMQTTProcessor: privateKey [%s]", *value);
+        securityPrivateKey_ = std::move(*value);
+        sslOpts_->privateKey = securityPrivateKey_.c_str();
       }
-      value = "";
-      if (context->getProperty(SecurityPrivateKeyPassWord.getName(), value) && !value.empty()) {
-        logger_->log_debug("AbstractMQTTProcessor: privateKeyPassword [%s]", value);
-        securityPrivateKeyPassWord_ = value;
-        sslopts_.privateKeyPassword = securityPrivateKeyPassWord_.c_str();
+      if (auto value = context->getProperty(SecurityPrivateKeyPassword)) {
+        logger_->log_debug("AbstractMQTTProcessor: privateKeyPassword [%s]", *value);
+        securityPrivateKeyPassword_ = std::move(*value);
+        sslOpts_->privateKeyPassword = securityPrivateKeyPassword_.c_str();
       }
     }
   }
+
+  if (auto last_will_topic = context->getProperty(LastWillTopic); last_will_topic.has_value() && !last_will_topic->empty()) {
+    last_will_ = MQTTAsync_willOptions_initializer;
+
+    logger_->log_debug("AbstractMQTTProcessor: Last Will Topic [%s]", *last_will_topic);
+    last_will_topic_ = std::move(*last_will_topic);
+    last_will_->topicName = last_will_topic_.c_str();
+
+    if (auto value = context->getProperty(LastWillMessage)) {
+      logger_->log_debug("AbstractMQTTProcessor: Last Will Message [%s]", *value);
+      last_will_message_ = std::move(*value);
+      last_will_->message = last_will_message_.c_str();
+    }
+
+    if (const auto value = context->getProperty<uint32_t>(LastWillQoS); value && (*value == MQTT_QOS_0 || *value == MQTT_QOS_1 || *value == MQTT_QOS_2)) {
+      logger_->log_debug("AbstractMQTTProcessor: Last Will QoS [%" PRIu32 "]", *value);
+      last_will_qos_ = {*value};
+      last_will_->qos = gsl::narrow<int>(last_will_qos_);
+    }
+
+    if (const auto value = context->getProperty<bool>(LastWillRetain)) {
+      logger_->log_debug("AbstractMQTTProcessor: Last Will Retain [%d]", *value);
+      last_will_retain_ = {*value};
+      last_will_->retained = last_will_retain_;
+    }
+  }
+
   if (!client_) {
-    MQTTClient_create(&client_, uri_.c_str(), clientID_.c_str(), MQTTCLIENT_PERSISTENCE_NONE, nullptr);
+    if (MQTTAsync_create(&client_, uri_.c_str(), clientID_.c_str(), MQTTCLIENT_PERSISTENCE_NONE, nullptr) != MQTTASYNC_SUCCESS) {
+      logger_->log_error("Creating MQTT client failed");
+    }
   }
   if (client_) {
-    MQTTClient_setCallbacks(client_, this, connectionLost, msgReceived, msgDelivered);
+    MQTTAsync_setCallbacks(client_, this, connectionLost, msgReceived, nullptr);
```

Comment:
should we check the return code for each MQTT function calls?

Reply:
What do you mean? Checking return value of `MQTTAsync_setCallbacks`?

Reply:
yes, I don't know which `MQTT*` functions return a "success/error", but this one seems to do so